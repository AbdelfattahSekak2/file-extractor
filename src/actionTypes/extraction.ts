export const GetExtraction = {
  FAILURE: "GET_EXTRACTION_FAILURE",
  SUCCESS: "GET_EXTRACTION_SUCCESS",
  REQUEST: "GET_EXTRACTION_REQUEST",
}

export const ClearExtractionState = {
  DEFAULT: "CLEAR_EXTRACTION_STATE_DEFAULT",
}
