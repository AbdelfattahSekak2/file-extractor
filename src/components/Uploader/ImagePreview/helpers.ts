import Extraction, { Prediction, Coordinates } from "../../../interfaces/extraction";

interface Shape extends Coordinates {
  strokeStyle: string;
}

const strokeStyles = ["red", "green", "blue"];


export const handleCanvas = (canvas: any, extraction: Extraction, file: File, setError: (error: string) => void) => {
  const image = canvas.loadImage(URL.createObjectURL(file));
  image.then((success: boolean) => {
    if (success) {
      canvas.setupCanvas();
      canvas.extractShapes(extraction);
      canvas.createShapes();
    } else {
      setError("Something went wrong loading the image");
    }
  });
}

export default class Canvas {
  canvasElement: HTMLCanvasElement;
  image: HTMLImageElement;
  context: CanvasRenderingContext2D | null;
  shapes: Shape[];

  constructor(canvasElement: HTMLCanvasElement) {
    this.canvasElement = canvasElement;
    this.image = new Image();
    this.context = canvasElement.getContext("2d");
    this.shapes = [];
  }

  public loadImage = async (file: string) => {
    return new Promise((resolve: (success: boolean) => void) => {
      this.image.onload = () => resolve(true);
      this.image.onerror = () => resolve(false);
      this.image.src = file;
    })
  };

  public setupCanvas = () => {
    const { width, height } = this.image;
    this.canvasElement.width = width;
    this.canvasElement.height = height;
    this.context!.drawImage(this.image, 0, 0, width, height);
  };

  public extractShapes = (extraction: Extraction) => {
    extraction.predictions.forEach((prediction: Prediction, index: number) => {
      Object.values(prediction).forEach((element: any | any[]) => {
        if (Array.isArray(element)) {
          element.forEach((attribute: any) => {
            if (attribute.hasOwnProperty("coord_x1")) {
              const { coord_x2, coord_x1, coord_y1, coord_y2 }: Coordinates = attribute;
              this.shapes.push({
                coord_x2,
                coord_x1,
                coord_y1,
                coord_y2,
                strokeStyle: strokeStyles[index],
              });
            };
          });
        } else {
          if (element.hasOwnProperty("coord_x1")) {
            const { coord_x2, coord_x1, coord_y1, coord_y2 }: Coordinates = element;
            this.shapes.push({
              coord_x2,
              coord_x1,
              coord_y1,
              coord_y2,
              strokeStyle: strokeStyles[index],
            });
          };
        };
      });
    });
  };

  public createShapes = () => this.shapes.forEach((shape: Shape) => {
    const { coord_x1, coord_x2, coord_y1, coord_y2, strokeStyle } = shape;
    const { width, height } = this.image;
    const x = width * coord_x1;
    const y = height * coord_y1;
    const rectWidth = (coord_x2 - coord_x1) * width;
    const rectHeight = (coord_y2 - coord_y1) * height
    this.context!.strokeStyle = strokeStyle;
    this.context!.strokeRect(x, y, rectWidth, rectHeight);
  })
}


