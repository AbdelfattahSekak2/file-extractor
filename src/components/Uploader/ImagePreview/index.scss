@import "../../../styles/mixins";

.image-preview {
  background: white;
  border-radius: $border-radius;
  height: 35vw;
  position: relative;
  canvas {
    border-radius: $border-radius;
    height: 100%;
    margin: auto;
  }
  i.icon {
    cursor: pointer;
    position: absolute;
    right: 2% !important;
    top: 2%;
    color: $dark-grey;
    &.supported {
      color: white !important;
    }
  }

  .not-supported {
    height: 100%;
    padding: 20px;
    width: 100%;
    display: flex;
    flex-direction: column;
    justify-content: center;
    align-items: center;
    p {
      font-size: 28px;
      font-weight: 500px;
      text-align: center;
      color: $dark-grey;
    }
    i.icon {
      color: $dark-grey;
      position: unset;
    }
  }
}

@media (max-width: 500px) {
  .image-preview {
    height: 100vw;
  }
}

@media (min-width: 500px) and (max-width: 1000px) {
  .image-preview {
    height: 60vw;
  }
}

@media (min-width: 1000px) and (max-width: 1500px) {
  .image-preview {
    height: 40vw;
  }
}
