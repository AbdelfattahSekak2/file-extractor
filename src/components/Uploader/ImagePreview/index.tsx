import React, { useEffect, useRef, Ref } from "react";
import cx from "classnames";
import { Icon } from "semantic-ui-react";

import "./index.scss";

import Extraction from "../../../interfaces/extraction";

import Canvas, { handleCanvas } from "./helpers";

interface IProps {
  file: File;
  handleDelete: () => void;
  extraction: Extraction;
  setError: (message: string) => void;
  supported: boolean
}

const ImagePreview: React.FunctionComponent<IProps> = ({
  handleDelete,
  file,
  extraction,
  setError,
  supported,
}) => {
  const canvasRef: Ref<HTMLCanvasElement> = useRef(null);

  useEffect(() => {
    if (
      extraction.predictions
      && extraction.predictions.length > 0
      && canvasRef.current
      && supported
    ) {
      const canvas = new Canvas(canvasRef.current);
      handleCanvas(canvas, extraction, file, setError);
    }
  }, [extraction, file, setError, supported])
  return <div className="image-preview">
    <Icon name="times" size="big" onClick={handleDelete} className={cx({ supported })} />
    {
      supported
        ? <canvas ref={canvasRef} id="canvas" />
        : <div className="not-supported">
          <Icon name="frown" size="huge" />
          <p>Sorry, we cannot render this file</p>
        </div>
    }
  </div>;
};

export default ImagePreview;

