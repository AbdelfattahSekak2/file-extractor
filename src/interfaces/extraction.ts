export interface Coordinates {
  coord_x1: number;
  coord_x2: number;
  coord_y1: number;
  coord_y2: number;
}

export interface Tax extends Coordinates {
  rate: number;
  amount: number;
  probability: number;
}

interface Date extends Coordinates {
  date_INTL: string,
  date_US: string,
  probability: 0.972,
}

interface Total extends Coordinates {
  amount: number,
  probability: number,
}

export interface Prediction {
  orientation: {
    degrees: string;
    probability: number;
  },
  category: {
    value: string;
    probability: number;
  },
  date: Date
  total: Total;
  taxes: Tax[]
}

export default interface Extraction {
  details: {
    call_id: string;
    computation_time: number;
  };
  predictions: Prediction[];
}
