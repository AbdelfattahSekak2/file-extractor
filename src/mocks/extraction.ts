import Extraction from "../interfaces/extractions";

export const extraction: Extraction = {
  "details": {
    "call_id": "49ef6b42-2724-4eaf-80bf-9add06a45176",
    "computation_time": 0.789,
  },
  "predictions": [
    {
      "orientation": {
        "degrees": "0",
        "probability": 0.723
      },
      "category": {
        "value": "parking",
        "probability": 0.302
      },
      "date": {
        "date_INTL": "2018-01-02",
        "date_US": "2012-10-26",
        "probability": 0.972,
        "coord_x1": 0.865,
        "coord_x2": 0.18,
        "coord_y1": 0.249,
        "coord_y2": 0.818
      },
      "total": {
        "amount": 0.732,
        "probability": 0.801,
        "coord_x1": 0.255,
        "coord_x2": 0.2,
        "coord_y1": 0.461,
        "coord_y2": 0.927,
      },
      "taxes": [
        {
          "rate": 0.476,
          "amount": 0.57,
          "probability": 0.647,
          "coord_x1": 0.216,
          "coord_x2": 0.256,
          "coord_y1": 0.16,
          "coord_y2": 0.145,
        }
      ]
    },
  ]
}
