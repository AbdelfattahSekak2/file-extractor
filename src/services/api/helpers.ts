export function identifyAction(action: any): string {
  return action.type.split("_").slice(0, -1).join("_");
}

export function getSuccessType(action: any): string {
  return `${identifyAction(action)}_SUCCESS`;
}

export function getFailType(action: any): string {
  return `${identifyAction(action)}_FAILURE`;
}

export function getRequestType(action: any): string {
  return `${identifyAction(action)}_REQUEST`;
}
