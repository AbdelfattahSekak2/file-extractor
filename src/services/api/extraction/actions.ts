import { ClearExtractionState, GetExtraction } from "../../../actionTypes/extraction";

export function getExtraction(file: FormData) {
  return {
    file,
    type: GetExtraction.REQUEST,
  };
}

export function clearExtractionState() {
  return {
    type: ClearExtractionState.DEFAULT,
  };
}
