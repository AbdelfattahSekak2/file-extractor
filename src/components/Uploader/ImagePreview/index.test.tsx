import React from "react";
import { ShallowWrapper, shallow, mount, ReactWrapper } from "enzyme";

import * as helpers from "./helpers";

import ImagePreview from ".";
import { extraction } from "../../../mocks/extraction";

describe("ImagePreview component", () => {
  // @ts-ignore
  helpers.handleCanvas = jest.fn();
  window.URL.createObjectURL = jest.fn(() => "url");
  const handleDelete = jest.fn();
  const file = new File([""], "filename", { type: 'text/html' });
  const setError = jest.fn();
  const defaultProps = {
    setError,
    handleDelete,
    file,
    extraction,
    supported: true,
  }
  let wrapper: ShallowWrapper | ReactWrapper;

  it("should render correctly", () => {
    wrapper = shallow(<ImagePreview {...defaultProps} />);
    expect(wrapper).not.toBeNull();
    expect(wrapper).toMatchSnapshot();
  });

  it("should handle file deletion correctly", () => {
    wrapper = mount(<ImagePreview {...defaultProps} />);
    expect(helpers.handleCanvas).toHaveBeenCalled();
  });

});
