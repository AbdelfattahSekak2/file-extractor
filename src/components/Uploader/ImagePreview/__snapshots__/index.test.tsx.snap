// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImagePreview component should render correctly 1`] = `
<div
  className="image-preview"
>
  <Icon
    as="i"
    className="supported"
    name="times"
    onClick={[MockFunction]}
    size="big"
  />
  <canvas
    id="canvas"
  />
</div>
`;
